dali/kernels/audio/mel_scale/mel_filter_bank_gpu.cu

```diff
@@ -124,51 +333,100 @@ class MelFilterBankGpu<T>::Impl : public MelFilterImplBase<T> {
       interval_ends_(args.nfilter + 2) {
     double mel = mel_low_ + mel_delta_;
     interval_ends_[0] = fftbin_start_;
-    interval_ends_[args.nfilter + 1] = fftbin_end_ + 1;
+    interval_ends_[args.nfilter + 1] = fftbin_end_;
+
     for (int interval = 1; interval < args_.nfilter + 1; interval++, mel += mel_delta_) {
       double freq = mel_scale.mel_to_hz(mel);
       interval_ends_[interval] = std::ceil(freq / hz_step_);
     }
+
+    weights_down_norm_.clear();
+    weights_down_norm_.resize(fftbin_size_);
+    weights_up_norm_.clear();
+    weights_up_norm_.resize(fftbin_size_);
+    fft2mel_.clear();
+    fft2mel_.resize(fftbin_size_, -1);
+
+    for (int f = fftbin_start_, interval = -1; f < fftbin_end_; f++) {
+      while (interval + 2 < static_cast<int>(interval_ends_.size()) &&
+             interval_ends_[interval + 2] <= f)
+        interval++;
+      assert(interval >= -1 && interval <= args.nfilter);
+      bool first = interval == -1;
+      bool last = interval == args.nfilter;
+
+      fft2mel_[f] = interval;
+
+      double w = weights_down_[f];
+      if (args.normalize) {
+        if (!first)
+          weights_down_norm_[f] = w * norm_factors_[interval];
+        if (!last)
+          weights_up_norm_[f] = (1 - w) * norm_factors_[interval + 1];
+      } else {
+        if (!first)
+          weights_down_norm_[f] = w;
+        if (!last)
+          weights_up_norm_[f] = (1 - w);
+      }
+    }
+    for (int f = fftbin_end_; f < fftbin_size_; f++)
+      fft2mel_[f] = args_.nfilter;
   }
 
   void Setup(ScratchpadEstimator &se, const TensorListShape<> &in_shape) {
-    se.add<mm::memory_kind::device, int>(interval_ends_.size());
-    se.add<mm::memory_kind::device, T>(weights_down_.size());
-    if (args_.normalize)
-      se.add<mm::memory_kind::device, T>(norm_factors_.size());
-
     inner_fft_ = true;
     for (int s = 0; s < in_shape.size(); s++) {
       inner_fft_ &= volume(in_shape.tensor_shape_span(s).begin() + args_.axis + 1,
                            in_shape.tensor_shape_span(s).end()) == 1;
     }
-    fft_dim_ = in_shape.tensor_shape_span(0)[args_.axis];
+    nfft_ = in_shape.tensor_shape_span(0)[args_.axis];
     if (inner_fft_) {
+      se.add<mm::memory_kind::device, int>(fft2mel_.size());
```

Comment:
This could be a great oportunity to get rid of ScratchpadEstimator usage. You might consider this out-of-scope though